### PR TITLE
MAINT: Add exception chaining where appropriate 

### PR DESCRIPTION
**Pull request by mdubravski**
This PR is related to #15986. Thanks!

### CODE REVIEW

File: numpy/core/code_generators/genapi.py [hunk 1]
```diff
@@ -281,9 +281,8 @@ def find_functions(filename, tag='API'):
                     state = SCANNING
                 else:
                     function_args.append(line)
-        except Exception:
-            print(filename, lineno + 1)
-            raise
+        except Exception as e:
+            raise Exception(filename, lineno + 1) from e
```
Comment (eric-wieser) [inline numpy/core/code_generators/genapi.py:285]: I think this should probably be:
```suggestion
        except ParseError:
            raise
        except Exception as e:
            raise ParseError(filename, lineno + 1, "see chained exception for details") from e
```

Comment (mdubravski) [reply numpy/core/code_generators/genapi.py]: Thank you for the insight-I just updated the code!

File: numpy/core/code_generators/genapi.py [hunk 2]
```diff
@@ -281,9 +281,10 @@ def find_functions(filename, tag='API'):
                     state = SCANNING
                 else:
                     function_args.append(line)
-        except Exception:
-            print(filename, lineno + 1)
+        except ParseError:
             raise
+        except Exception as e:
+            raise ParseError(filename, lineno + 1, "see chained exception for details") from e
```
Comment (charris) [inline numpy/core/code_generators/genapi.py:287]: Lint is complaining that this line is too long, line lengths need to be < 80 characters.  This is a bit strange as there may be two ParseError exceptions.

Comment (mdubravski) [reply numpy/core/code_generators/genapi.py:291]: I see, I can shorten the line and push the changes.

**Comment by charris** [comment]
close/reopen

**Comment by charris** [comment]
Thanks @mdubravski .

File: numpy/core/code_generators/genapi.py [hunk 3]
```diff
@@ -281,9 +281,11 @@ def find_functions(filename, tag='API'):
                     state = SCANNING
                 else:
                     function_args.append(line)
-        except Exception:
-            print(filename, lineno + 1)
+        except ParseError:
```
Comment (thegautamkumarjaiswal) [inline numpy/core/code_generators/genapi.py:284]: improved

**Review by ghost — COMMENTED** [review]
parse error will help to find excetion more related to this error. nice

**Pull request merged**